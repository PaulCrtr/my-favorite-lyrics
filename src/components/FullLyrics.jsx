import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import addItem from "./addItem";
import deleteItem from "./deleteItem";
import iconAdd from "../img/icon-add.png";
import trash from "../img/trashdelete.png";
import "../style/fullLyrics.css";

const FullLyrics = (props) => {
  const { datas } = props.location.state;
  const [alreadyAdded, setAlreadyAdded] = useState(false);
  const [isClicked, setIsClicked] = useState(false);

  // Check if the item is already added
  useEffect(() => {
    if (datas) {
      const searchItem = JSON.parse(
        window.localStorage.getItem("myLyrics")
      ).find((element) => element.id === datas.id);
      if (searchItem) {
        setAlreadyAdded(true);
      } else {
        setAlreadyAdded(false);
      }
    }
  }, [datas, isClicked]);

  return (
    <>
      {datas && datas && (
        <div className="page-lyrics">
          <div className="lyrics-header">
            <Link to="/">Home</Link>
            <Link to="/list">My list</Link>
            <div>
              {alreadyAdded ? (
                <img
                  src={trash}
                  alt="trash"
                  onClick={() => {
                    setIsClicked(!isClicked);
                    deleteItem(datas.id);
                  }}
                />
              ) : (
                <img
                  src={iconAdd}
                  alt="add lyrics"
                  onClick={() => {
                    setIsClicked(!isClicked);
                    addItem(datas);
                  }}
                />
              )}
            </div>
          </div>
          <iframe
            title={datas.title}
            scrolling="no"
            frameBorder="0"
            allowtransparency="true"
            src={`https://www.deezer.com/plugins/player?format=classic&autoplay=false&playlist=true&width=700&height=350&color=007FEB&layout=dark&size=medium&type=tracks&id=${datas.id}&app_id=1`}
            width="300"
            height="90"
          />

          <h2>{`${datas.artist} - ${datas.title}`}</h2>
          <p className="lyrics">{datas.lyrics}</p>
        </div>
      )}
    </>
  );
};

export default FullLyrics;
