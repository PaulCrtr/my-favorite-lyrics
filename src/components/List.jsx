import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import deleteItem from "./deleteItem";
import groupList from "./groupList";

const List = () => {
  const [isClicked, setIsClicked] = useState(false);
  const [list, setList] = useState(groupList());

  useEffect(() => {
    setList(groupList());
  }, [isClicked]);

  return (
    <div>
      <Link to="/">Home</Link>
      {Object.keys(list).map((key, i) => (
        <div key={i}>
          <h2>{key}</h2>
          {list[key].map((datas) => (
            <div key={datas.id}>
              <Link to={{ pathname: "/lyrics", state: { datas } }}>
                <h3>{datas.title}</h3>
                <img src={datas.picture} alt={datas.artist} />
              </Link>
              <button
                onClick={() => {
                  setIsClicked(!isClicked);
                  deleteItem(datas.id);
                }}
              >
                Remove lyrics
              </button>
            </div>
          ))}
        </div>
      ))}
    </div>
  );
};

export default List;
