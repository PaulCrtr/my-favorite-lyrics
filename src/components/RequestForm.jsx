import React, { useState, useEffect } from "react";
import { Redirect } from "react-router-dom";
import axios from "axios";

const RequestForm = ({ setItem, item, setIsLoading }) => {
  const [search, setSearch] = useState("");
  const [results, setResults] = useState([]);
  const [isTyping, setIsTyping] = useState(false);
  const [typingTime, setTypingTime] = useState(0);

  // Set a timeout during writing so that the API is not overloaded
  useEffect(() => {
    setIsTyping(true);
    if (typingTime) {
      clearTimeout(typingTime);
    }
    setTypingTime(
      setTimeout(() => {
        setIsTyping(false);
      }, 300)
    );
  }, [search]);

  // Get search suggests and related datas
  useEffect(() => {
    if (!isTyping && search)
      axios
        .get(`https://api.lyrics.ovh/suggest/${search}`)
        .then((res) => {
          let tempSearch = [];
          for (let i = 0; i < 3; i++) {
            const datas = res.data.data[i];
            if (datas) {
              tempSearch.push({
                id: datas.id,
                artist: datas.artist.name,
                title: datas.title,
                picture: datas.artist.picture,
              });
            }
          }
          setResults(tempSearch);
        })
        .catch(() => setResults([]));
  }, [isTyping]);

  // Get lyrics and pass them into the item object
  const getlyrics = (i) => {
    setIsLoading(true);
    axios
      .get(`https://api.lyrics.ovh/v1/${results[i].artist}/${results[i].title}`)
      .then((res) => res.data.lyrics)
      .then((res) => {
        setIsLoading(false);
        setItem({ ...results[i], lyrics: res });
      })
      .catch(() => {
        setIsLoading(false);
        setItem("notFound");
      });
  };

  return (
    <div className="request-form">
      {item && item.lyrics && (
        <Redirect
          to={{
            pathname: "/lyrics",
            state: { datas: item },
          }}
        />
      )}
      <div className="input-container">
        <input
          type="text"
          value={search}
          onChange={(e) => setSearch(e.target.value)}
          placeholder="Search"
        />
      </div>
      <ul className="suggests">
        {search &&
          results.map((element, i) => (
            <li
              key={element.id}
              id={i}
              onClick={(e) => {
                getlyrics(e.currentTarget.id);
              }}
            >
              <p>{`${element.artist} - ${element.title}`}</p>
              <img src={element.picture} alt={element.artist} />
            </li>
          ))}
      </ul>
    </div>
  );
};

export default RequestForm;
