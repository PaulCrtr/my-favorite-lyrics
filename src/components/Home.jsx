import React, { useState, useEffect } from "react";
import RequestForm from "./RequestForm";

const Home = () => {
  const [item, setItem] = useState();

  return (
    <div>
      <RequestForm setItem={setItem} />
      <p>{item && item.lyrics}</p>
    </div>
  );
};

export default Home;
