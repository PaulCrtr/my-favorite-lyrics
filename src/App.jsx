import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import Home from "./components/Home";
import List from "./components/List";
import FullLyrics from "./components/FullLyrics";

const App = () => {
  return (
    <Router>
      <Switch>
        <Route path="/" exact component={Home} />
        <Route path="/list" exact component={List} />
        <Route path="/lyrics" exact component={FullLyrics} />
      </Switch>
    </Router>
  );
};

export default App;
