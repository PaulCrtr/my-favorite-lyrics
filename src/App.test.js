import React from "react";
import { shallow } from "enzyme";
import App from "./App";
import Home from "./components/Home";
import List from "./components/List";
import FullLyrics from "./components/FullLyrics";
import RequestForm from "./components/RequestForm";

const testItem1 = {
  id: 10,
  artist: "test 1",
  title: "test 1",
  picture: "test 1",
  lyrics: "This is a test 1.",
};

const testItem2 = {
  id: 20,
  artist: "test 2",
  title: "test 2",
  picture: "test 2",
  lyrics: "This is a test 2.",
};

describe("App", () => {
  beforeAll(() => {
    localStorage.myLyrics = JSON.stringify([testItem1, testItem2]);
  });

  it("renders App correctly", () => {
    const appWrapper = shallow(<App />);
    expect(appWrapper).toMatchSnapshot();
  });

  it("renders Home correctly", () => {
    const homeWrapper = shallow(<Home />);
    expect(homeWrapper).toMatchSnapshot();
  });

  it("renders RequestForm correctly", () => {
    const requestFormWrapper = shallow(<RequestForm />);
    expect(requestFormWrapper).toMatchSnapshot();
  });

  it("renders List correctly", () => {
    const listWrapper = shallow(<List />);
    expect(listWrapper).toMatchSnapshot();
  });

  it("renders Lyrics correctly", () => {
    const fullLyricsWrapper = shallow(
      <FullLyrics location={{ state: { datas: testItem1 } }} />
    );
    expect(fullLyricsWrapper).toMatchSnapshot();
  });
});
