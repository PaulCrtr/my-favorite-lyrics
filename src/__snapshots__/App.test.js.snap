// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App correctly 1`] = `
<BrowserRouter>
  <Switch>
    <Route
      component={[Function]}
      exact={true}
      path="/"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/list"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/lyrics"
    />
  </Switch>
</BrowserRouter>
`;

exports[`App renders Home correctly 1`] = `
<div
  className="home"
>
  <div
    className="home-nav"
  >
    <h1>
      Lyrics
    </h1>
    <Link
      to="./list"
    >
      My list
    </Link>
  </div>
  <RequestForm
    setIsLoading={[Function]}
    setItem={[Function]}
  />
</div>
`;

exports[`App renders List correctly 1`] = `
<div
  className="list"
>
  <div
    className="home-link-container"
  >
    <Link
      className="home-link"
      to="/"
    >
      Home
    </Link>
  </div>
  <div
    className="artist"
    key="0"
  >
    <div
      className="artist-header"
    >
      <h2>
        test 1
      </h2>
      <img
        alt="test 1"
        src="test 1"
      />
    </div>
    <div
      className="track"
      key="10"
    >
      <Link
        to={
          Object {
            "pathname": "/lyrics",
            "state": Object {
              "datas": Object {
                "artist": "test 1",
                "id": 10,
                "lyrics": "This is a test 1.",
                "picture": "test 1",
                "title": "test 1",
              },
            },
          }
        }
      >
        <h3>
          test 1
        </h3>
      </Link>
      <img
        alt="delete"
        onClick={[Function]}
        src="trashdelete.png"
      />
    </div>
  </div>
  <div
    className="artist"
    key="1"
  >
    <div
      className="artist-header"
    >
      <h2>
        test 2
      </h2>
      <img
        alt="test 2"
        src="test 2"
      />
    </div>
    <div
      className="track"
      key="20"
    >
      <Link
        to={
          Object {
            "pathname": "/lyrics",
            "state": Object {
              "datas": Object {
                "artist": "test 2",
                "id": 20,
                "lyrics": "This is a test 2.",
                "picture": "test 2",
                "title": "test 2",
              },
            },
          }
        }
      >
        <h3>
          test 2
        </h3>
      </Link>
      <img
        alt="delete"
        onClick={[Function]}
        src="trashdelete.png"
      />
    </div>
  </div>
</div>
`;

exports[`App renders Lyrics correctly 1`] = `
<Fragment>
  <div
    className="page-lyrics"
  >
    <div
      className="lyrics-header"
    >
      <Link
        to="/"
      >
        Home
      </Link>
      <Link
        to="/list"
      >
        My list
      </Link>
      <div>
        <img
          alt="add lyrics"
          onClick={[Function]}
          src="icon-add.png"
        />
      </div>
    </div>
    <iframe
      allowtransparency="true"
      frameBorder="0"
      height="90"
      scrolling="no"
      src="https://www.deezer.com/plugins/player?format=classic&autoplay=false&playlist=true&width=700&height=350&color=007FEB&layout=dark&size=medium&type=tracks&id=10&app_id=1"
      title="test 1"
      width="300"
    />
    <h2>
      test 1 - test 1
    </h2>
    <p
      className="lyrics"
    >
      This is a test 1.
    </p>
  </div>
</Fragment>
`;

exports[`App renders RequestForm correctly 1`] = `
<div
  className="request-form"
>
  <div
    className="input-container"
  >
    <input
      onChange={[Function]}
      placeholder="Search songs"
      type="text"
      value=""
    />
  </div>
  <ul
    className="suggests"
  />
</div>
`;
